### ISSUE TITLE
missing `Content-Length` header

### ISSUE THREAD

**Issue by oyvindberg** [http4s/http4s-grpc#154]
I've been wondering for the longest time why this library doesn't work with most clients.

i finally figured it out now. In `ClientGrpc.{unaryToUnary, unaryToStream}` a body is computed, but since it's streaming, the library doesn't stop to consider how long it is. When the request is passed to the underlying client library, the body is simply dropped because of it. 

see for instance `org.http4s.jdkhttpclient.JdkHttpClient`:
```scala
          val rb = HttpRequest.newBuilder
            .method(
              req.method.name,
              if (req.isChunked)
                BodyPublishers.fromPublisher(publisher)
              else
                req.contentLength match {
                  case Some(length) if length > 0L =>
                    BodyPublishers.fromPublisher(publisher, length)
                  case _ => BodyPublishers.noBody
                }
            )
```

This is at least the case for ember and jdk clients. The netty client did work, and is likely the odd one out without this particular check.

one possible fix is something like this:
```scala
  def unaryToUnary[F[_]: Concurrent, A, B]( // Stuff We can provide via codegen
      encode: Encoder[A],
      decode: Decoder[B],
      serviceName: String,
      methodName: String
  )( // Stuff We can apply at application scope
      client: Client[F],
      baseUri: Uri
  )( // Stuff we apply at invocation
      message: A,
      ctx: Headers
  ): F[B] =
    codecs.Messages.encodeSingle(encode)(message).compile.toVector.flatMap { bytes =>
      val req = Request[F](Method.POST, baseUri / serviceName / methodName, HttpVersion.`HTTP/2`)
        .putHeaders(
          SharedGrpc.TE,
          SharedGrpc.GrpcEncoding,
          SharedGrpc.GrpcAcceptEncoding,
          SharedGrpc.ContentType,
          `Content-Length`(bytes.size.toLong)
        )
        .putHeaders(ctx.headers.map(Header.ToRaw.rawToRaw)*)
        .withBodyStream(Stream.emits(bytes))
        .withAttribute(H2Keys.Http2PriorKnowledge, ())

      client
        .run(req)
        .use(resp =>
          handleFailure(resp.headers) >>
            codecs.Messages
              .decodeSingle(decode)(resp.body)
              .handleErrorWith(e => resp.trailerHeaders.flatMap(handleFailure[F]).attempt.flatMap(t => t.as(e).merge.raiseError[F, B])) <*
            resp.trailerHeaders.flatMap(handleFailure[F])
        )
    }
```

though you may want to use chunks maybe to keep the properties of how it works now.

this is likely the cause of #122 . the consequence of not sending that body is that a successful response will come back with no body, which again breaks some assumptions, and ultimately will leave ember for instance hanging.

**Comment by hamnis** [comment]
This is not the cause of #122. That happens because of ember failed to close the stream when told so by the trailers. This has been fixed in http4s-ember-client 0.23.30. That said, we can include something like this, with encoding to a single chunk. I have a preliminary PR on that here #156

**Comment by danicheg** [comment]
IIUC, there is no strict requirement for clients to have `Content-Length` sent in the gRPC specification. I quickly checked Ember's H2 client, and it doesn't appear to rely on that header [when sending data](https://github.com/http4s/http4s/blob/v0.23.30/ember-core/shared/src/main/scala/org/http4s/ember/core/h2/H2Stream.scala#L78). @oyvindberg so the root cause for your potential issue with the Ember is something else.

**Comment by danicheg** [comment]
I also find it odd that the `JdkHttpClient` decides on the presence of the request's body by looking at the `Content-Length` (I haven't checked this precisely, speaking as per @oyvindberg's assumption). Perhaps that part should be aligned with Ember's approach. @hamnis wdyt?

**Comment by oyvindberg** [comment]
@hamnis maybe not. The symptoms caused by this problem appeared similar (weird errors like references to `GoAway`, stack traces pointing to body/trailer handling code, ultimately a hanging state). After applying this fix I can run a bunch of requests using for instance ember client without the fix for #122. 

>  @oyvindberg so the root cause for your potential issue with the Ember is something else.

So I went digging into the source of the jdk client, found this issue, hacked together a fix, and saw that the same fix also made my code work with ember client. I didn't go into ember itself to see.

I suppose the logical thing to do is to write a unit test which threads data through `ClientGrpc` with different clients and assert that the resulting request have body.

**Comment by hamnis** [comment]
> I also find it odd that the `JdkHttpClient` _solely_ decides on the presence of the request's body by looking at the `Content-Length` (I haven't checked this precisely, speaking as per @oyvindberg's assumption). Perhaps that part should be aligned with Ember's approach. @hamnis wdyt?
> https://github.com/http4s/http4s/blob/v0.23.30/ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala#L135-L139

Yes. Since we are doing a 0.10 series release of jdk http client we should do that now

**Comment by hamnis** [comment]
https://github.com/http4s/http4s-jdk-http-client/pull/1146
